### PR TITLE
4 garden metadata creation

### LINKED ISSUE #4: Garden metadata creation
_As a Garden Publisher I want to create garden metadata so consumers can easily find my model_

# Assumptions
1. Metadata will include:
 a. Model title
 b. Model description
 c. DataCite information about the submission
2. Data will be defined using [pydantic](https://pydantic-docs.helpmanual.io)
3. The GardenClient interface will expose a [fluent API](https://en.wikipedia.org/wiki/Fluent_interface) starting with a `create_model` method.
4. This story ends with the library writing a JSON file containing this metadata to the user's current working directory 

#Acceptance Criteria
Given I have instantiated a `GardenClient` when I call the `create_garden` with chained metadata setting methods specifying all of the required fields then I will find a JSON metadata file in my current working directory.

Given I have instantiated a `GardenClient` when I call the `create_garden` with chained metadata setting methods, committing a required field then I will receive an exception.

### PR DESCRIPTION
**Pull request by OwenPriceSkelly**
(Tentatively) this closes #4 

----
:books: Documentation preview :books:: https://garden-ai--21.org.readthedocs.build/en/21/

### CODE REVIEW

File: garden_ai/garden.py [hunk 1]
```diff
@@ -79,3 +81,65 @@ def _authenticate(self):
             on_refresh=self.auth_key_store.on_refresh,
         )
         return authorizer
+
+    def create_garden(self, **kwargs):
```
Comment (BenGalewsky) [inline garden_ai/garden.py:85]: My vision was to make the required fields named parameters in this constructor so it's clear in your IDE exactly what is the minimum for creating a Garden.

File: garden_ai/garden.py [hunk 2]
```diff
@@ -79,3 +81,65 @@ def _authenticate(self):
...
+
+            gc = GardenClient()
+            pea_garden = gc.create_garden(
+                creators=["Mendel, Gregor"], title="Experiments on Plant Hybridization"
+            )
+            pea_garden.year = 1863
+        """
+        return Garden(**kwargs)
+
+    def register_garden(self, G: Garden):
```
Comment (BenGalewsky) [inline garden_ai/garden.py:113]: Capital letter for this parameter seems non-pythonic, right?

File: garden_ai/model.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+from datetime import datetime
+from enum import Enum, auto
+from typing import List
+from uuid import uuid4, UUID
+import pydantic
+from pydantic import BaseModel, Field, validator
+
+
+class TempField(Enum):
```
Comment (BenGalewsky) [inline garden_ai/model.py:13]: This seems complicated... what is going on here that's not just built into pydantic?

File: garden_ai/garden.py [hunk 4]
```diff
@@ -79,3 +81,65 @@ def _authenticate(self):
...
+        def request_doi(garden):
+            # TODO
+            # to be replaced with code to build the actual doi
+            return "10.55555/fake-doi"
+        try:
+            # attempts to populate a dummy garden with the data from the real
+            # one, so that it invokes our custom validators on every field (in
+            # particular, those which still have TempField defaults)
+            G.doi = request_doi(G)
+            _ = Garden(**G.dict())
```
Comment (BenGalewsky) [inline garden_ai/garden.py:137]: Might was well make `validate` a first class method on `Garden` - others may want to call this directly

File: pyproject.toml [hunk 5]
```diff
@@ -28,6 +29,16 @@ safety = "^2.3.1"
 sphinx = "4.3.2"
 sphinx-rtd-theme = "^1.0.0"
 
+
+[tool.poetry.group.dev]
+optional=true
+
+[tool.poetry.group.dev.dependencies]
+python-lsp-server = {extras = ["rope"], version = "^1.6.0"}
```
Comment (BenGalewsky) [inline pyproject.toml:37]: Hmm, does anyone else use these dependencies? I wonder how the team feels about adding such specific development tooling to our common pyproject?

**Comment by BenGalewsky** [comment]
Overall this is a really great start! You've made a very abstract set of requirements concrete. 
I've added some specific comments. More generally, you will need to add unit tests to keep our [code coverage](https://app.codecov.io/gh/Garden-AI/garden/tree/4-garden-metadata-creation) up.

File: garden_ai/garden.py (hunk 1 above)
Comment (BenGalewsky) [inline garden_ai/garden.py:85]: Also wonder if this should just be a Garden constructor - not sure if having this as a method on GardenClient adds anything

File: garden_ai/garden.py (hunk 2 above)
Comment (ascourtas) [reply garden_ai/garden.py:113]: agreed, also full-named parameters are preferred to abbreviations.

File: garden_ai/garden.py [hunk 6]
```diff
@@ -79,3 +81,65 @@ def _authenticate(self):
             on_refresh=self.auth_key_store.on_refresh,
         )
         return authorizer
+
+    def create_garden(self, **kwargs):
+        """Initialize and return a new Garden object
```
Comment (ascourtas) [inline garden_ai/garden.py:86]: some more clarity in the docstring on what a Garden object is would be helpful

File: garden_ai/garden.py [hunk 7]
```diff
@@ -79,3 +81,65 @@ def _authenticate(self):
...
+
+        If invoked with keyword arguments corresponding to field names (e.g.
```
Comment (ascourtas) [inline garden_ai/garden.py:88]: this sentence is a little confusing, please clarify

File: garden_ai/garden.py [hunk 8]
```diff
@@ -79,3 +81,65 @@ def _authenticate(self):
...
+        `title` or `creator`) are passed directly to the `Garden` model, making
+        it possible for the user to provide necessary info upfront to be parsed
+        and validated at initialization, or to incrementally provide field
+        values via simple assignment to attributes of the returned object
+        directly.
+
+        Parameters
+        ----------
+        **kwargs :
+            Metadata for the new Garden object. Keyword arguments matching
```
Comment (ascourtas) [inline garden_ai/garden.py:98]: how will a user know what these kwargs should be? please add more context

File: garden_ai/garden.py [hunk 9]
```diff
@@ -79,3 +81,65 @@ def _authenticate(self):
...
+            required or recommended fields will be coerced to the appropriate
+            type, if necessary.
+
+        Examples
```
Comment (ascourtas) [inline garden_ai/garden.py:102]: love the example! <3

File: garden_ai/garden.py [hunk 10]
```diff
@@ -79,3 +81,65 @@ def _authenticate(self):
...
+            return "10.55555/fake-doi"
+        try:
+            # attempts to populate a dummy garden with the data from the real
+            # one, so that it invokes our custom validators on every field (in
+            # particular, those which still have TempField defaults)
+            G.doi = request_doi(G)
+            _ = Garden(**G.dict())
+        except ValidationError as e:
+            logger.error(str(e))
+            print(e)
```
Comment (ascourtas) [inline garden_ai/garden.py:140]: please remove print statements

File: garden_ai/garden.py (hunk 4 above)
Comment (ascourtas) [inline garden_ai/garden.py:137]: please add an inline comment about what is being accomplished here. the existing `Garden` metadata is being unpacked to create a new `Garden` object? feels funny

File: garden_ai/garden.py [hunk 11]
```diff
@@ -79,3 +81,65 @@ def _authenticate(self):
...
+            gc = GardenClient()
+            pea_garden = gc.create_garden(
+                creators=["Mendel, Gregor"], title="Experiments on Plant Hybridization"
+            )
+            pea_garden.year = 1863
+        """
+        return Garden(**kwargs)
+
+    def register_garden(self, G: Garden):
+        """Finalize the setup of a Garden object with user-populated fields
```
Comment (ascourtas) [inline garden_ai/garden.py:114]: please add more context. What does "finalize" mean? The method name has the word "register" -- what is being registered here? from the code, it looks like the key component is the minting of the DOI. Please make that clear in the docstring

File: garden_ai/garden.py (hunk 1 above)
Comment (OwenPriceSkelly) [reply garden_ai/garden.py:85]: I like that idea -- we could also just expose the `Garden` constructor to the user, because my impression is that pydantic makes a point of playing nicely with IDEs/etc

File: garden_ai/garden.py [hunk 12]
```diff
@@ -79,3 +81,65 @@ def _authenticate(self):
...
+        try:
+            # attempts to populate a dummy garden with the data from the real
+            # one, so that it invokes our custom validators on every field (in
+            # particular, those which still have TempField defaults)
+            G.doi = request_doi(G)
+            _ = Garden(**G.dict())
+        except ValidationError as e:
+            logger.error(str(e))
+            print(e)
+            raise TypeError
```
Comment (ascourtas) [inline garden_ai/garden.py:141]: if you are raising a TypeError, I suggest wrapping it with some context so the user sees the logged error more easily. Something like the following, but with more context (hard for me to tell from this method):

```
raise TypeError(f"Error while registering Garden: {e.message}")
```

File: garden_ai/garden.py (hunk 1 above)
Comment (OwenPriceSkelly) [reply garden_ai/garden.py:85]: Oh, but that would be providing completion suggestions for all the fields, not *only* the required fields, which I think is what you were getting at

File: garden_ai/garden.py (hunk 2 above)
Comment (OwenPriceSkelly) [reply garden_ai/garden.py:113]: indeed

File: garden_ai/garden.py [hunk 13]
```diff
@@ -79,3 +81,65 @@ def _authenticate(self):
...
+        else:
+            outfile = Path.cwd() / 'metadata.json'
```
Comment (ascourtas) [inline garden_ai/garden.py:143]: could be good to log to `info` that a file is being written locally?

Comment (ascourtas) [inline garden_ai/garden.py:143]: not immediately clear what is happening in this line, please add inline comment

File: garden_ai/model.py [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+    REQUIRED = auto()  # just those needed to get a doi
+    WARN = auto()
+    # setting this should trigger warning before publishing the model
+    # we can also inspect the object to see which fields have been set manually
+    # might be more flexible in the future
+
+
+class Garden(BaseModel):
+    """Object representation of a Garden™.
```
Comment (ascourtas) [inline garden_ai/model.py:22]: lol the little TM ✨

File: garden_ai/garden.py (hunk 1 above)
Comment (BenGalewsky) [reply garden_ai/garden.py:85]: It would be something like `create_garden(title: str, authors: list[str], **kwargs)` 

So the mandatory fields would be obvious, and the rest would be keywords

File: garden_ai/model.py [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+        other Gardeners.
+
+
+    Optional Attributes
+    ----------
+    doi: str
+        A garden's doi can be manually set by the user, or generated automatically via the
+        DataCite rest api using the required fields.
+
+    language: str
```
Comment (ascourtas) [inline garden_ai/model.py:50]: Nice, but add clarity on what this means -- I assume it's just a metadata field explaining what language the abstract, etc are written in?

Otherwise, it could be read as a field for setting the language (ie we do translation)

File: garden_ai/model.py (hunk 14 above)
Comment (ascourtas) [reply garden_ai/model.py:22]: Please add a little context about what a Garden is

File: garden_ai/model.py [hunk 16]
```diff
@@ -0,0 +1,146 @@
...
+
+    subjects: List[str]
+        Subjects, keywords, classification codes, or key phrases pertaining to the Garden.
+
+    Examples
+    --------
+    Creating a new Garden with a ``GardenClient`` instance::
+
+        gc = GardenClient()
+        pea_garden = gc.create_garden(
```
Comment (ascourtas) [inline garden_ai/model.py:62]: <3

File: garden_ai/model.py [hunk 17]
```diff
@@ -0,0 +1,146 @@
...
+                                  for Big Pea Data extraction and classification.
+                                  It consists of a 2-hectare plot behind the monastery,
+                                  and a 30,000-plant dataset.'''
+
+        gc.register(pea_garden) # we can also have this invoke a pydantic "root validator" we define elsewhere
+
+    Notes
+    --------
+    Mendel's work was ignored by the scientific community during his lifetime,
+    presumably due to the lack of a working DOI.
```
Comment (ascourtas) [inline garden_ai/model.py:77]: 😂 amazing. High quality work.

File: garden_ai/model.py [hunk 18]
```diff
@@ -0,0 +1,146 @@
...
+        validate_assignment = True  # validators invoked (only?) on assignment
+
+    creators: List[str] = Field(TempField.REQUIRED, min_items=1, unique_items=True)
+    title: str = Field(TempField.REQUIRED)
+    resourceTypeGeneral: str = "Other"  # (or: model, software, service, interactive?)
+    publisher: str = "Garden"
+    year: str = Field(default_factory=lambda: str(datetime.now().year))
+
+    # dois currently only validated against this regex
+    # doi: str = Field(TempField.REQUIRED, alias="identifier", regex=r"^10.\d{4,9}/[-._;()/:A-Z0-9]+$")
```
Comment (ascourtas) [inline garden_ai/model.py:100]: is the commented out code in this Config meant to be here?

File: poetry.lock [hunk 19]
```diff
@@ -31,6 +31,29 @@ python-versions = ">=3.6"
 [package.dependencies]
 pytz = ">=2015.7"
 
+[[package]]
```
Comment (ascourtas) [inline poetry.lock:34]: are you supposed to commit the lock file? This was contentious for Pipenv in its time

**Review by ascourtas — COMMENTED** [review]
Great job Owen! This is good quality code. I especially loved the examples. I added some comments on structure and readability/understandability. 

I didn't comment on the Pydantic class defs as much because I am less familiar -- I will leave that for @WardLT ✨

File: garden_ai/garden.py (hunk 1 above)
Comment (ascourtas) [reply garden_ai/garden.py:85]: I was also thinking this was a little funny, but Owen's example made the logical user flow really nice? I am torn.

**Review by WardLT — COMMENTED** [review]
I like the mechanism for marking recommended fields, but think we should have it occur when a model is being exported (e.g., via `dict()` or `json()`) so that users don't get warnings that their object is incomplete before finishing. 

Maybe write "check recommended fields" as a function and overload the `dict`/`json` methods to run it first before export. 

We could also look up the field description in the `__fields__` attribute of the class so that we can print out a "what is this field" in the warning. 

How does that sound?

File: pyproject.toml (hunk 5 above)
Comment (OwenPriceSkelly) [reply pyproject.toml:37]: I was also conflicted about this, the lsp tooling I use was deeply confused by poetry and this seemed to be the workaround -- do you know if there's a less obnoxious way for me to include those as dev dependencies? I could also just not commit those changes to pyproject.toml on my end, but that feels like asking for trouble

File: garden_ai/garden.py (hunk 1 above)
Comment (OwenPriceSkelly) [reply garden_ai/garden.py:85]: I agree with both points, I think @BenGalewsky's vision earlier about `create_garden(title: str, authors: list[str], **kwargs)` is a nice compromise

File: garden_ai/model.py (hunk 3 above)
Comment (BenGalewsky) [reply garden_ai/model.py:13]: I wonder if DOI will be set in the client or added in the server anyway. If that's the case then we have our 100% required fields, which are set in the constructor.. everything else is 100% optional, so we don't need any fancy footwork.

Even if the DOI is fetched in the client, then it's not technically a required field. We would just go set it if it's empty and carry on. I think we building extra complexity that is not actually required

File: garden_ai/garden.py (hunk 4 above)
Comment (OwenPriceSkelly) [reply garden_ai/garden.py:137]: yeah I agree it does feel funny. that's what the few lines of comments above are failing  to totally explain -- the way the pydantic model is currently set up, we're allowed to build a blank object that has invalid default values (taken from the `TempField` enum in model.py) but only checks for validity when something tries to explicitly set them. 

So if someone has a Garden they _think_ they've finished but they never gave a title, this would look to pydantic like someone tried to construct a dummy Garden that explicitly sets `title=TempField.REQUIRED`, which would implicitly trigger whatever helpful pydantic errors/exceptions we want it to

tldr: one-liner that forces a call to every validator, in case there are still fields that haven't been validated yet

File: garden_ai/garden.py (hunk 11 above)
Comment (OwenPriceSkelly) [reply garden_ai/garden.py:114]: this is the most placeholder-y part of the code I think. my reasoning here was (copied from my convo with Logan in slack): if a user is invoking some `GardenClient` method to finalize/register their `Garden` object, that's a safe point to assume the user is "done" (or thinks they're done) providing metadata and seemed like a reasonable point to run a full suite of validation functions

File: garden_ai/garden.py (hunk 6 above)
Comment (ascourtas) [reply garden_ai/garden.py:86]: That is an excellent question that I think is also tied to the above comment about whether or not `create_garden()` should just be a regular constructor for a class. 

In the past, we have used names like `FoundryMetadata` or `FoundryBaseModel` to describe Pydantic classes describing the metadata, where then `Foundry` might be the class that describes the actual Foundry object. 

I am not as familiar with Pydantic, so my suggestion would be to read into what the best practices are for how to conceptually organize this to be in-line with OOP principles. 

A major question here is we know that the Pydantic class describes the state (ie attributes) of a Garden, but what about the function? Does that go in another `Garden` class, or just in the `GardenClient`? 

If there are no clear answers, I'd lean towards the Pydantic class being called `GardenMetadata` or something similar, and then having the actual object (which contains `GardenMetadata` as an attribute) be `Garden` or `GardenClient`, depending on where we want function to live.

File: garden_ai/garden.py (hunk 11 above)
Comment (OwenPriceSkelly) [reply garden_ai/garden.py:114]: I think that design-wise this method is probably trying to do too much, and to be fully transparent I was kinda hoping "register" would be evocative but vague enough to not think too hard about what it meant

File: garden_ai/model.py (hunk 18 above)
Comment (OwenPriceSkelly) [reply garden_ai/model.py:100]: the regex convenience option that pydantic provided there was causing later validation for the doi to skip for some reason, I'm not sure if it was a pydantic bug or a me bug and I thought I might revisit it later, but also "I'll revisit this later" could be famous last words so I can just take it out

File: garden_ai/garden.py (hunk 10 above)
Comment (OwenPriceSkelly) [reply garden_ai/garden.py:140]: oh I left this one in on purpose but I meant to ask if it should be a log instead -- this print statement was there to display the pydantic Helpful Error Message to the user the way they do in the pydantic docs examples

File: garden_ai/model.py [hunk 20]
```diff
@@ -0,0 +1,146 @@
...
+    garden_id: UUID = Field(default_factory=uuid4, allow_mutation=False)
+
+    # always=False should prevent this from being called until an attempt to set
+    # a field, but this should be called for every field.
+    @validator("*", pre=True, always=False)
+    def check_required(cls, val, field: pydantic.fields.ModelField):
+        assert val is not TempField.REQUIRED, f"{field.name} must be set"
+        if val is TempField.WARN:
+            # TODO: log warning
+            print(f"Warning: {field.name} is recommended but unset")
```
Comment (Aadit-Ambadkar) [inline garden_ai/model.py:118]: Wouldn't it make more sense to do logging instead of print statements?

File: garden_ai/garden.py [hunk 21]
```diff
@@ -79,3 +81,65 @@ def _authenticate(self):
...
+        TypeError
+
+        Examples
+        --------
+        TODO
+        """
+        def request_doi(garden):
+            # TODO
+            # to be replaced with code to build the actual doi
+            return "10.55555/fake-doi"
```
Comment (Aadit-Ambadkar) [inline garden_ai/garden.py:131]: Consider making this a publicly available function. The user may want access to this.

File: garden_ai/garden.py (hunk 10 above)
Comment (Aadit-Ambadkar) [reply garden_ai/garden.py:140]: Use logging instead of print statements.

File: garden_ai/garden.py (hunk 12 above)
Comment (Aadit-Ambadkar) [reply garden_ai/garden.py:141]: 👍🏼 agreed

File: garden_ai/garden.py [hunk 22]
```diff
@@ -79,3 +81,65 @@ def _authenticate(self):
...
+            G.doi = request_doi(G)
+            _ = Garden(**G.dict())
+        except ValidationError as e:
+            logger.error(str(e))
+            print(e)
+            raise TypeError
+        else:
+            outfile = Path.cwd() / 'metadata.json'
+            with open(outfile, 'w+') as f:
+                f.write(G.json())
```
Comment (Aadit-Ambadkar) [inline garden_ai/garden.py:145]: what's the purpose of this? (this is a geniuine question, not a review comment)

File: garden_ai/model.py (hunk 20 above)
Comment (blaiszik) [reply garden_ai/model.py:118]: yes, good idea Aadit!

File: garden_ai/garden.py (hunk 6 above)
Comment (OwenPriceSkelly) [reply garden_ai/garden.py]: Context so I can resolve this in good conscience: the current `Garden` object is on track to be greater than the sum of its metadata, with more functionality than would be provided by a more vanilla pydantic model. This feels ok to me, but I do think you're right here to question whether this is good OOP, & my feelings won't be hurt if we later decide that the interface would be better if garden metadata and ~gardening functionality were owned by different classes

File: garden_ai/garden.py (hunk 11 above)
Comment (OwenPriceSkelly) [reply garden_ai/garden.py]: Made it clearer in docstring that this is mainly for tying up loose metadata-ends

File: garden_ai/garden.py (hunk 21 above)
Comment (OwenPriceSkelly) [reply garden_ai/garden.py]: agreed, ended up attaching the method to the Garden object, but might make sense in the future to move it to the garden client object

File: garden_ai/garden.py (hunk 22 above)
Comment (OwenPriceSkelly) [reply garden_ai/garden.py]: just to hit the acceptance criteria spelled out in #4
